pygbm/splitting.py

```diff
@@ -60,78 +61,163 @@ def __init__(self, n_features, binned_features, n_bins,
         else:
             self.constant_hessian_value = float32(1.)  # won't be used anyway
 
-    def split_indices(self, sample_indices, split_info):
-        binned_feature = self.binned_features.T[split_info.feature_idx]
-        left_idx, right_idx = 0, 0
-        sample_indices_left = np.empty_like(sample_indices)
-        sample_indices_right = np.empty_like(sample_indices)
-        # TODO: parallelize this loop by blocks (e.g. 1e6-sized chunks)?
-        # That would involve some overhead with concatenating the blocks at
-        # the end. Alternatively we could make sample_indices a list of numpy
-        # arrays in the grower tree node datastructure to ease
-        # parallelization.
-        for sample_idx in sample_indices:
+
+@njit(parallel=True,
+      locals={'sample_idx': uint32,
+              'left_count': uint32,
+              'right_count': uint32})
+def split_indices(context, sample_indices, split_info):
+    """Split samples into left and right arrays."""
+
+    # This is a multi-threaded implementation inspired from lightgbm.
+    # Quick break-down:
+    # - samples indices is devided into n_threads regions. Each thread will
+    #   be responsible for its own region. Here is an example with 4
+    #   threads and 24 samples:
+    #   samples_indices = [abcdef|ghijkl|mnopqr|stuvwx]
+    # - Each thread processes 6 = 24 // 4 entries and maps them
+    #   into tmp_left or tmp_right. For example,
+    #   we could have the following mapping:
+    #   - tmp_left =  [abef..|il....|mnopqr|tux...]
+    #   - tmp_right = [cd....|ghjk..|......|svw...]
+    # - We keep track of the start positions of the regions (the '|') as
+    #   well as the size of each region. We also keep track of the number
+    #   of samples put into the left/right node by each thread. Concretely:
+    #   - left_cnt =  [4, 2, 6, 3]
+    #   - right_cnt = [2, 4, 0, 3]
+    # - Finally, we put tmp_left and tmp_right into the returned arrays
+    #   sample_indices_left and sample_indices_right. This is done with the
+    #   help of left_write_offset and right_write_offset, whose entries are
+    #   here denoted by '^':
+    #   - sample_indices_left =   [abefilmnopqrtux.........]
+    #                              ^   ^ ^     ^
+    #   - samples_indices_right = [cdghjksvw...............]
+    #                              ^ ^   ^
+    #                                    ^
+
+    # Note: we could get rid of the tmp_xxxx arrays in our case and
+    # directly work on samples_indices_left/right. I'm leaving them
```

Comment:
Have you tried to use samples_indices_left / samples_indices_right directly instead of using the `tmp_*` arrays? I suspect this would run faster as you don't need to copy the indices at the end of this function (the last `prange` loop).

Reply:
I can try but I don't think we'll gain much, the second prange loop is still needed because there is no way to know in advance how many samples land into the left or right node, for each thread. So we would gain the array allocation time only.

Reply:
Indeed. Let's keep this way then. But remove this comment then.